File: python/mag_memory/harness.py
"""
Minimal AgentMemoryHarness interface.

This is the Python-side counterpart to the Rust ``AgentMemoryHarness`` trait.
It provides three operations — *observe*, *retrieve_context*, *trace* —
without prescribing how they are implemented underneath.
"""

from __future__ import annotations

import json
import os
import time
import uuid
from abc import ABC, abstractmethod
from pathlib import Path
from typing import Any, Dict, Optional

from .mcp_client import McpClient


class AgentTurn:
    """
    A single turn in an agent trajectory.

    This is intentionally unstructured — callers pass whatever fields
    are relevant for their domain.
    """

    def __init__(
        self,
        turn_idx: int,
        action: str = "",
        observation: str = "",
        metadata: Optional[Dict[str, Any]] = None,
    ):
        self.turn_idx = turn_idx
        self.action = action
        self.observation = observation
        self.metadata = metadata or {}

    def to_text(self) -> str:
        """Serialise to the AMA-Bench trajectory text format."""
        parts = [f"Step {self.turn_idx}:"]
        if self.action:
            parts.append(f"Action: {self.action}")
        if self.observation:
            parts.append(f"Observation: {self.observation}")
        return "\n".join(parts)


class AgentMemoryHarness(ABC):
    """
    Three-method harness that sits between a benchmark and MAG.

    Implementations may use the MCP stdio interface (for Python callers)
    or the Rust storage traits directly (for in-process callers).
    """

    @abstractmethod
    def observe(self, turn: AgentTurn) -> None:
        """Ingest a single turn into memory."""
        ...

    @abstractmethod
    def retrieve_context(self, query: str) -> str:
        """Retrieve relevant context for *query*."""
        ...

    @abstractmethod
    def trace(self, event: Dict[str, Any]) -> None:
        """Emit an unstructured JSON trace event."""
        ...


class MagHarness(AgentMemoryHarness):
    """
    In-process harness backed by a :class:`McpClient`.

    This is suitable for Python benchmarks that want to drive MAG
    turn-by-turn rather than batch-store an entire trajectory.
    """

    def __init__(
        self,
        session_id: Optional[str] = None,
        project: Optional[str] = None,
        trace_dir: Optional[str] = None,
        mag_binary: Optional[str] = None,
        home_dir: Optional[str] = None,
        search_mode: str = "text",
        search_limit: int = 10,
    ):
        self.session_id = session_id or str(uuid.uuid4())
        self.project = project or f"harness_{self.session_id[:8]}"
        self.search_mode = search_mode
        self.search_limit = search_limit

        self._trace_dir = Path(trace_dir or os.path.expanduser("~/.mag/traces"))
        self._trace_dir.mkdir(parents=True, exist_ok=True)
        self._trace_file = self._trace_dir / f"{self.session_id}.jsonl"

        self._client = McpClient(
            mag_binary=mag_binary,
            home_dir=home_dir,
            timeout_secs=120.0,
        )
        self._client.start()

    def observe(self, turn: AgentTurn) -> None:
        content = turn.to_text()
        self._client.store(
            content=content,
            tags=["harness", "turn"],
            project=self.project,
            session_id=self.session_id,
            metadata={
                "turn_idx": turn.turn_idx,
                **turn.metadata,
            },
        )
        self.trace(
            {
                "event": "observe",
                "turn_idx": turn.turn_idx,
                "content_chars": len(content),
            }
        )

    def retrieve_context(self, query: str) -> str:
        result = self._client.search(
            query=query,
            mode=self.search_mode,
            limit=self.search_limit,
            advanced=True,
            project=self.project,
        )
        results = result.get("results", [])
        context = "\n\n".join(r.get("content", "") for r in results)
        self.trace(
            {
                "event": "retrieve",
                "query": query,
                "results": len(results),
                "context_chars": len(context),
            }
        )
        return context

    def trace(self, event: Dict[str, Any]) -> None:
        event["_ts"] = time.time()
        event["_session"] = self.session_id
        with open(self._trace_file, "a", encoding="utf-8") as f:
            f.write(json.dumps(event, ensure_ascii=False) + "\n")

    def close(self) -> None:
        self._client.stop()

File: src/harness.rs
//! AgentMemoryHarness — minimal three-method interface between agents and MAG.
//!
//! This trait is intentionally small (observe / retrieve_context / trace) so
//! that benchmarks and agent loops can swap storage implementations without
//! changing their orchestration code.
//!
//! # Phase 2a scope
//!
//! - Trait definition only.
//! - No in-process implementations yet (those ship with Phase 2b MemoryArena
//!   integration).
//! - The AMA-Bench adapter uses the Python-side harness via MCP stdio.

use anyhow::Result;
use serde_json::Value;

/// A single turn in an agent trajectory.
///
/// All fields are optional — callers pass whatever is relevant for their
/// domain (software engineering, web navigation, embodied AI, etc.).
#[derive(Debug, Clone, Default)]
pub struct AgentTurn {
    /// Turn index within the episode.
    pub turn_idx: usize,
    /// Agent action (e.g. tool call, code edit, navigation command).
    pub action: String,
    /// Environment observation (e.g. tool output, test result, page HTML).
    pub observation: String,
    /// Unstructured metadata (timestamps, tool names, file paths, etc.).
    pub metadata: Value,
}

impl AgentTurn {
    /// Build a turn from action + observation text.
    pub fn new(turn_idx: usize, action: impl Into<String>, observation: impl Into<String>) -> Self {
        Self {
            turn_idx,
            action: action.into(),
            observation: observation.into(),
            metadata: Value::Null,
        }
    }

    /// Attach metadata and return self for chaining.
    pub fn with_metadata(mut self, metadata: Value) -> Self {
        self.metadata = metadata;
        self
    }

    /// Serialise to the AMA-Bench trajectory text format.
    pub fn to_text(&self) -> String {
        let mut parts = vec![format!("Step {}:", self.turn_idx)];
        if !self.action.is_empty() {
            parts.push(format!("Action: {}", self.action));
        }
        if !self.observation.is_empty() {
            parts.push(format!("Observation: {}", self.observation));
        }
        parts.join("\n")
    }
}

/// Minimal harness between an agent loop and a memory backend.
///
/// Implementations may use in-process storage (e.g. `SqliteStorage`) or
/// communicate with a remote MAG server via MCP.
pub trait AgentMemoryHarness: Send + Sync {
    /// Ingest a single turn into memory.
    fn observe(&self, turn: &AgentTurn) -> Result<()>;

    /// Retrieve relevant context for a query.
    ///
    /// Callers compose the returned string into their prompts — the harness
    /// does not perform injection itself.
    fn retrieve_context(&self, query: &str) -> Result<String>;

    /// Emit an unstructured JSON trace event.
    ///
    /// Events are appended to `~/.mag/traces/<run-id>.jsonl` by typical
    /// implementations, but the trait does not prescribe a sink.
    fn trace(&self, event: Value) -> Result<()>;
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn agent_turn_to_text() {
        let turn = AgentTurn::new(3, "Edit src/main.rs", "Added println! for debugging");
        let text = turn.to_text();
        assert!(text.contains("Step 3:"));
        assert!(text.contains("Action: Edit src/main.rs"));
        assert!(text.contains("Observation: Added println! for debugging"));
    }

    #[test]
    fn agent_turn_empty_fields_omitted() {
        let turn = AgentTurn::new(0, "", "hello");
        let text = turn.to_text();
        assert!(text.contains("Step 0:"));
        assert!(!text.contains("Action:"));
        assert!(text.contains("Observation: hello"));
    }
}
